### ISSUE TITLE
ASYNCIFY=1 does not work with functions that need to return a value back to JS

### ISSUE THREAD

**Issue by juj** [emscripten-core/emscripten#20492]
ASYNCIFY=1 does not work with functions that need to return a value back to JS side, for example `emscripten_request_animation_frame_loop()`.

`a.cpp`

```c++
#include <emscripten.h>
#include <emscripten/html5.h>

extern "C" void wgpu_buffer_map_sync(int buffer, int mode, double offset, double size);

EM_BOOL raf(double time, void *userData)
{
  EM_ASM(console.log(`rAF at time ${$0}`), time);

  // Map a GPU buffer synchronously
  wgpu_buffer_map_sync(1,2,3,4);

  // Return EM_TRUE to signal we want to keep the rAF loop going on.
  return EM_TRUE;
}

int main()
{
  emscripten_request_animation_frame_loop(raf, 0);
}
```

`lib_a.js`

```js
mergeInto(LibraryManager.library, {
  wgpu_buffer_map_sync__sig: 'viidd',
  wgpu_buffer_map_sync__async: true,
  wgpu_buffer_map_sync: (opts) => Asyncify.handleAsync(async () => {
    await 42;
  }),
});
```

```
em++ a.cpp --js-library lib_a.js -sASYNCIFY=1 -o a.html
```
The intent with this program is to have the application render continuously in a rAF loop, and map a WebGPU buffer in each tick of the loop.

However, `wgpu_buffer_map_sync()` ends up returning integer `0` back to caller, and the frame loop will not be set up to continue.

(With brief testing, it seems that maybe `-sASYNCIFY=2` does not have this issue)

I wonder if there is any way to fix the above issue?

**Comment by kripken** [comment]
> However, wgpu_buffer_map_sync() ends up returning integer 0 back to caller, and the frame loop will not be set up to continue.

That function doesn't return a value (it even has a signature starting with `v`, void) - am I reading your comment wrong somehow?

In general I think this should work, and we should have tests for it. cc @brendandahl who might have a better memory though

**Comment by sbc100** [comment]
I think the problem is that the call to `wgpu_buffer_map_sync` can/well cause the `ref` function to unwind the wasm stack all the way out to the RAF code in `test/test_browser.py` which does not know how to deal with unwound stacks and will interpret the nullish return value of the unwond stack to be a signal to top the RAF loop.

In other words, I don't think the RAF loop code in `test/test_browser.py` can deal with RAF callbacks that suspend themselves.

Also, it seems like a very odd thing for a RAF callback to do... RAF callbacks are supposed to render to the screen and then another is not supposed to run until the next frame.. if one RAF callback syspends.. what happens if it resumes N frames later.. that would be a little odd wouldn't it?

**Comment by kripken** [comment]
Thanks @sbc100 , now I see. That all sounds right.

If we want this to work, I think we'd need to update the RAF loop code to check for an unwind and handle that.

**Comment by juj** [comment]
> Also, it seems like a very odd thing for a RAF callback to do... RAF callbacks are supposed to render to the screen and then another is not supposed to run until the next frame.. if one RAF callback syspends.. what happens if it resumes N frames later.. that would be a little odd wouldn't it?

No, this is how the WebGPU authors are proposing that would work to enable emulating synchronous GPU map buffer access.

>If we want this to work, I think we'd need to update the RAF loop code to check for an unwind and handle that.

It would be nice to do this. How should an unwind be detected?

**Comment by kripken** [comment]
> How should an unwind be detected?

I think we can check for `Asyncify.state == Asyncify.State.Unwinding`. Though that gets unset when the unwind ends, which I'm not sure if it happens before or after this JS would be reached - I suspect before, since it happens on the boundary of the wasm, and this is the JS that called into the wasm? If that doesn't work, `Asyncify.currData` will be non-null during a sleep.

**Comment by RReverser** [comment]
> I think we can check for `Asyncify.state == Asyncify.State.Unwinding`. Though that gets unset when the unwind ends, which I'm not sure if it happens before or after this JS would be reached - I suspect before, since it happens on the boundary of the wasm, and this is the JS that called into the wasm? If that doesn't work, `Asyncify.currData` will be non-null during a sleep.

I believe other places use combination of `Asyncify.currData` to detect sleeping operation and `Asyncify.whenDone` to actually wait for it.

**Comment by RReverser** [comment]
FWIW when I had similar usecase with Asyncify I tend to just avoid `emscripten_request_animation_frame_loop` and implement it via Asyncify too with something like:

```js
EM_ASYNC_JS(void, request_animation_frame, (), {
  await new Promise(requestAnimationFrame);
});

int main() {
  for (;;) {
    // ...do something and then
    request_animation_frame();
  }
}
```

If it's possible to do transparently, maybe `emscripten_request_animation_frame_loop` could similarly switch to different implementation when Asyncify is enabled?
